Add "und" diacritics (#28)

Mottie opened the pull request:
Added "und" (undetermined) diacritics extracted from mark.js.

The `build-undetermined.js` file only needs to be run once and after the diacritics list is updated. It doesn't really need to be included, but I left it in just-in-case.

Mottie committed:
Add "und" diacritics (#28)

julkue commented on build/build-undetermined.js:
```diff
@@ -0,0 +1,54 @@
+  // This file only needs to be run once, or if the diacritics list is updated!
+  // copied from https://git.io/vNPua
+  diacritics = [
+    'aàáảãạăằắẳẵặâầấẩẫậäåāą', 'AÀÁẢÃẠĂẰẮẲẴẶÂẦẤẨẪẬÄÅĀĄ',
```
What will happen if we have diacritics that are included in the "und" and another language? For example I guess since you've added the PL language to the validated list, this diacritics list contains them too.

Mottie replied:
We have characters defined in multiple languages already. If you want to build the array for mark.js, then only target the `und` language.

julkue replied:
So they are generated redundant?
Using `und` isn't future proof. For me this is a temporary workaround until we've classified the diacritics and created the corresponding language files. Then, `und` will not be completed or empty at all.

julkue replied:
@Mottie So, if I would roll out mark.js with the "und" language I could not only use this undetermined language since I can not "pin" the output to a specific "und" state so once the diacritics inside there would be converted to a different language I'd loose them. Our versioned database & API only covers schema changes, not data changes, which I like since usually this means that you automatically receive updates.

So I'd need to use all languages. The question then is, if they are generated redundant? So I created an entry in "de.json" for "à" and tested it. It seems to be redundantly generated. So for the "und" case in my opinion it's necessary to check if the diacritic is included in another language and if so, ignore it and print a warning to the console. This should be part of the generation process of the entire build, not the generation process of und.json since this may only be important in future once other guys create language files.

What do you think? If you agree, can you take over this part or should I?

julkue commented on build/build-undetermined.js:
```diff
@@ -0,0 +1,54 @@
+    'zžżź', 'ZŽŻŹ'
+  ];
+
+function buildEntries(block) {
```
Please use an arrow function instead.

julkue commented on build/templates/und.json:
```diff
@@ -0,0 +1,21 @@
+{
+  "metadata": {
+    "alphabet": "Zzzz",
```
Is this valid?

Mottie replied:
Yes. See http://www.unicode.org/reports/tr35/#Unknown_or_Invalid_Identifiers.

Mottie committed:
Use arrow function

julkue reviewed (changes requested):
We need to discuss the topic "redundant data".

Mottie committed:
Only include diacritics not found in the database

Mottie commented:
The "und" entry now only contains diacritics not found in any other language.

julkue commented:
@Mottie I'm not sure whether we understood each other correctly. 

> This should be part of the generation process of the entire build, not the generation process of und.json since this may only be important in future once other guys create language files.

What I meant is this:

1. We create the und.json automatically. Now with your changes applied we check if the diacritics are included in another language and only copy them over to und.json when not.
2. In future, we're moving parts of the diacritics of "und.json" into a language file
3. What then happens: They are either generated redundantly or you need to run the "und" generation process each time you change something in the database

Therefore I said that to make the process easier in future, the check should be part of the general build, not the und build. However, the alternative now is to trigger the und build in the common build task. Choose what you prefer

Mottie commented:
Currently, it's not ideal... every time a new language is added or updated, this would need to be done:

* Run `npm run build` to update the `diacritics.json` file.
* Run `npm run build:und` to update the `und.json`.
* Run `npm run build` again to update the `und` section.

I was just thinking that the `build-undetermined.js` file would be a run one-time script that builds the "und.json" file. Then in the `build.js` file, the "und" entry would be built first and as each diacritic is encountered, it is removed from the "und" entry - the build script would only need to be run once.

julkue commented:
I can follow you until here:

> Then in the build.js file, the "und" entry would be built first

Not sure about the part afterwards. But what I think you mean is what I mean: The common build generates the diacritics.json and only includes those diacritic characters from the "und" language that aren't included in other language files.

Mottie commented:
Yes, that is what I was trying to explain.

julkue commented:
@Mottie Then I agree

Mottie committed:
Remove "und" duplicates in the build

Mottie committed:
Fix codeclimate issue

Mottie commented:
Ok, so now with this set up...

* Use `npm run build:und` only if the internal list of diacritics is updated. It builds an `und.json` file with *all* of the listed diacritics.
* Use `npm run build` to build the diacritics database file. Within the process, duplicate entries within the `und` entry will be removed.

julkue reviewed (changes requested):
1. The process correctly determines diacritics included in language files and skips them in the "und" language. However, there's no console warning for the case that a diacritics inside the und.json is already contained in another language file. This should be a user friendly warning to allow editor themself to update the file
2. When you delete the folder "und" for testing purposes and run the und build you get an error: `Error writing und.json { Error: ENOENT: no such file or directory, open 'P:\open-source\diacritics-database\src\und\und.json'`
3. The indentation of und.json is invalid. Use two spaces like in all other json files or use .jsbeautifyrc directly as a config.

Mottie committed:
Add build cleanup of und

Mottie committed:
Build run; undetermined file cleanup

Mottie committed:
Manually remove undetermined data

The "l" and "L" represent the base character which is no longer needed
since the associated diacritics were removed

Mottie commented:
All fixed & ready for review.

I added a `build-cleanup.js` file to remove the duplicate diacritics from the `build-undetermined.js` and `und.json` files automatically; but you will notice though that the last commit requires the manual removal of the base character - `'lł', 'LŁ'` became `'l', 'L'` and is no longer needed, so I had to manually remove it.

julkue commented on build/build-cleanup.js:
```diff
@@ -0,0 +1,72 @@
+/*!***************************************************
```
Thanks! But please move the file into `processes/`

julkue commented on build/build-cleanup.js:
```diff
@@ -0,0 +1,72 @@
+   * Save modified "build-undetermined.js" and "und/und.json" files
+   */
+  saveFiles() {
+    fs.writeFile(this._buildFile, this._buildContent, 'utf8', err => {
```
Once we have merged #30 (first) then we can use the write method from there.

julkue commented on build/build.js:
```diff
@@ -24,6 +25,7 @@ class Build {
    * Constructor
    */
   constructor() {
+    this._diacritics = new Set([]);
```
Why prefixed by `_`? We don't have privaty properties

julkue commented:
I'm a bit confused. Change the diacritics inside `build-undetermined.js` to:

```js
    this.diacritics = [
      'aàảãạăằắẳẵặầấẩẫậåā', 'AÀẢÃẠĂẰẮẲẴẶẦẤẨẪẬÅĀ',
      'cč', 'CČ',
      'dđď', 'DĐĎ',
      'eèẻẽẹêềếểễệëěē', 'EÈẺẼẸÊỀẾỂỄỆËĚĒ',
      'iìỉĩịī', 'IÌỈĨỊÏĪ',
      'nň', 'NŇ',
      'oòỏõọôồốổỗộơởỡớờợøō', 'OÒỎÕỌÔỒỐỔỖỘƠỞỠỚỜỢØŌ',
      'rř', 'RŘ',
      'sšș', 'SŠȘ',
      'tťțţ', 'TŤȚŢ',
      'üuùủũụưừứửữựůū', 'UÙỦŨỤƯỪỨỬỮỰŮŪ',
      'yỳỷỹỵÿ', 'YỲỶỸỴŸ',
      'zž', 'ZŽ'
    ];
```

And run the und build. You'll notice that many diacritics will have the base value of "ü". From my point of view no diacritics are allowed in base. So, I'm curious what's happening here?

What I would expect is that the build determines that "ü" is already included in e.g. "de" and therefore prints a warning to the console and skips the diacritic.

Mottie committed:
Change "capital" to "case"

Mottie committed:
Move & tweak build-cleanup.js

Mottie committed:
Remove underscore from diacritics variable

Mottie commented:
The `this.diacritics` value is following the convention used in mark.js. The first character is the base and all following diacritics will be associated with that base. I guess it really wasn't a convention, but that's how it was set up.

Should we account for the base being elsewhere in the string? I don't think it's necessary since this is a file that is only required to be run after a new diacritic is added to the list.

Mottie committed:
Fix paths in build-cleanup

Mottie committed:
Fix file name in cleanup

julkue commented:
@Mottie Could you please not revert things inside a PR? PR's are going to get squash-merged, therefore it's unnecessary and prevents me from understanding the actual changes.

![topic](https://user-images.githubusercontent.com/8310677/35904694-6333ae40-0be4-11e8-9d15-e13c458f5251.png)

> The this.diacritics value is following the convention used in mark.js. 

Which convention do you mean? As far as I can see there's only the convention to prefix the variable with an underscore when it is used with getters/setters (e.g. `get X()`).

> Should we account for the base being elsewhere in the string? I don't think it's necessary since this is a file that is only required to be run after a new diacritic is added to the list.

I'm sorry, I don't understand what you mean.

julkue commented on build/build-undetermined.js:
```diff
@@ -0,0 +1,104 @@
+    const base = block.shift(),
+      caseValue = Utils.getCase(base);
+    block.forEach(diacritic => {
+      und.data[diacritic] = {
```
Please don't manipulate variables directly from outside the class. Create a clone in the class, e.g.:

```js
constructor(submission = und) {
   this.und = {...submission};
}
```

Then you can use `this.und` in further calls.

julkue commented on build/build-undetermined.js:
```diff
@@ -0,0 +1,104 @@
+    if (!fs.existsSync(path)) {
+      fs.mkdirSync(path);
+    }
+    fs.writeFile(
```
Please use the utils method

julkue commented on build/processes/build-cleanup.js:
```diff
@@ -0,0 +1,64 @@
+      if (output.und.und.data[char]) {
+        delete output.und.und.data[char];
+        delete und.data[char];
+        this.buildContent = this.buildContent.replace(char, '');
```
`this.buildContent` seems not to be declared in the constructor. Please always declare variables in the constructor, otherwise you don't know what's stored inthe class without digging into each method.

julkue reviewed (changes requested):
As mentioned above I have the problem that the build doesn't detect redundant diacritics between "und" and other languages. There's no warning or removal. Also the file will generate `base` values of a diacritic (see example above). No clue what's happening here.

Mottie committed:
Use Utils to write JSON

Mottie committed:
Manipulate template clone

Mottie committed:
Cleanup: Load js file in constructor

Mottie commented:
> PR's are going to get squash-merged

I rebased this branch with the master and there were lots of conflicts, especially in the `build.js`. It would have been messy either way.

> Which convention do you mean?

For example, if we look at the ["c" entry](https://github.com/julmot/mark.js/blob/master/src/lib/regexpcreator.js#L292) from mark.js: `'cçćč'`

It has the base character (`'c'`) followed by three associated diacriticis (`'çćč'`): `'{base}{diacriticsx3}'`

The `build-undetermined.js` file assumes this and uses `split('')` then `shift()` to get the first (base) character. I was saying I didn't think it was necessary to account for the base being elsewhere in the string, e.g. `'{diacritic}{base}{diacritics...}'` since the data will be edited by us (and the cleanup script) and the file only needs to be run if and when a new diacritic is added to the list.

julkue commented:
Thanks @Mottie.

> I was saying I didn't think it was necessary to account for the base being elsewhere in the string, 

I agree. And I also do understand what you mean now.

However, my question is still open. If you put in the following in build-undetermined.js:

```js
    this.diacritics = [
      'aàảãạăằắẳẵặầấẩẫậåā', 'AÀẢÃẠĂẰẮẲẴẶẦẤẨẪẬÅĀ',
      'cč', 'CČ',
      'dđď', 'DĐĎ',
      'eèẻẽẹêềếểễệëěē', 'EÈẺẼẸÊỀẾỂỄỆËĚĒ',
      'iìỉĩịī', 'IÌỈĨỊÏĪ',
      'nň', 'NŇ',
      'oòỏõọôồốổỗộơởỡớờợøō', 'OÒỎÕỌÔỒỐỔỖỘƠỞỠỚỜỢØŌ',
      'rř', 'RŘ',
      'sšș', 'SŠȘ',
      'tťțţ', 'TŤȚŢ',
      'uùủũụưừứửữựůūü', 'UÙỦŨỤƯỪỨỬỮỰŮŪÜ',
      'yỳỷỹỵÿ', 'YỲỶỸỴŸ',
      'zž', 'ZŽ'
    ];
```

(NOTE: An `ü` and `Ü` was added to the end of the array entries which is contained in at least DE)

then I would expect that there is a warning log in the console saying that "Ü" is already contained in another language and also is removed from the output. Both is not the case. Am I misunderstanding something?

Mottie committed:
Remove file before saving

Mottie commented:
Oh, the `writeJSON` script wouldn't save the changes because it checks if the file exists first. Fixed now.

julkue commented on build/build-undetermined.js:
```diff
@@ -0,0 +1,92 @@
+    Utils.writeJSON(
+      './src/und/',
+      'und',
+      JSON.stringify(this.und, null, 2) + '\n'
```
The pretty print thing should be on the side of the utils method.

julkue committed:
Remove redundant json pretty print

julkue committed:
Don't directly manipulate outer variables

julkue committed:
Fix missing JSDoc comments

julkue commented on src/mr/mr.json:
```diff
@@ -17,6 +17,7 @@
   },
   "data": {
     "ऱ": {
+      "case": "none",
```
Not sure why this wasn't generated before.

julkue commented:
🚀 Thanks @Mottie 🚀
